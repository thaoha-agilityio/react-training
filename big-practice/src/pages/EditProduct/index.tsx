import { memo, useCallback } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { Box, Spinner } from '@chakra-ui/react';

// Components
import Banner from '@components/Banner';
import Form from '@components/Form';
import Container from '@components/Container';

// Constants
import { DATA_CRUMBS, ROUTES, SUCCESS_MESSAGES } from '@constants';

// Hooks
import { useFetchProductDetail, useMutationEditProduct } from '@hooks';

// Types
import { IProduct, STATUSES } from '@types';
import { useCustomToast } from '@hooks';

const EditProduct = (): JSX.Element => {
  const navigate = useNavigate();
  const { uuid = '' } = useParams();
  const { showToast } = useCustomToast();

  // Get the mutate from useMutationEditProduct hook
  const { mutate, isLoading } = useMutationEditProduct();

  // Handle the submission of the product form
  const handleAddProduct = useCallback(
    (value: IProduct) => {
      const { id, name, image, price } = value;
      return mutate(
        {
          ...value,
          id,
          name,
          image,
          price,
        },
        {
          onError: (error) => {
            showToast(STATUSES.ERROR, error.message);
          },
          onSuccess: (res) => {
            navigate(ROUTES.DETAIL_PRODUCT_PARAMS + res.id);
            showToast(STATUSES.SUCCESS, SUCCESS_MESSAGES.EDITED(name));
          },
        },
      );
    },
    [mutate],
  );

  const { data: product, isFetching } = useFetchProductDetail(uuid);

  return (
    <>
      <Banner title='Edit Product' crumbs={DATA_CRUMBS} />
      <Container>
        <Box as='main' pt='30px'>
          {isFetching ? (
            <Box textAlign='center'>
              <Spinner variant='secondary' />
            </Box>
          ) : (
            <Form
              title='Edit Product'
              product={product}
              onSubmitProduct={handleAddProduct}
              isLoading={isLoading}
            />
          )}
        </Box>
      </Container>
    </>
  );
};

export default memo(EditProduct);
