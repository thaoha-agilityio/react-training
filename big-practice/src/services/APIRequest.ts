import axios, { AxiosError, AxiosResponse } from 'axios';

export const getData = async <T>(url: string): Promise<T> => {
  try {
    const res: AxiosResponse<T> = await axios.get<T>(url);

    return res.data;
  } catch (error) {
    throw error as AxiosError;
  }
};

export const postData = async <T>({ item, url }: { item: T; url: string }): Promise<T> => {
  try {
    const response = await axios.post<T>(url, item);

    return response.data;
  } catch (error) {
    throw error as AxiosError;
  }
};

export const putData = async <T>({ item, url }: { item: Partial<T>; url: string }): Promise<T> => {
  try {
    const response = await axios.put<T>(url, item);

    return response.data;
  } catch (error) {
    throw error as AxiosError;
  }
};

export const deleteData = async <T>(url: string): Promise<T> => {
  try {
    const response = await axios.delete<T>(url);

    return response.data;
  } catch (error) {
    throw error as AxiosError;
  }
};

const api = { getData, postData, putData };

export { api };
