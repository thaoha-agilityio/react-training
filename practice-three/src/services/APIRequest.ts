import axios from "axios";

const getData = async <T>(url: string): Promise<T | any> => {
  try {
    const res = await axios.get<T>(url);

    return {
      data: res.data,
      error: "",
    };
  } catch (error: any) {
    return {
      data: [],
      error: error,
    };
  }
};

const api = { getData };

export { api };
