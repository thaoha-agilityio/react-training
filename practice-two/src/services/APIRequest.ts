import axios from './config';

const getData = <T>(url: string, params?: object) => axios.get<never, T>(url, params);

export { getData };
