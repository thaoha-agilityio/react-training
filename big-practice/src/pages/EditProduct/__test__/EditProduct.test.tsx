import { fireEvent, act } from '@testing-library/react';
// Helpers
import { renderWithRouterAndQuery } from '@helpers';

// Hooks
import * as hooks from '@hooks';

// Service
import { api } from '@services/APIRequest';

// Constants
import { MOCK_PRODUCTS } from '@constants';

import EditProduct from '..';

// Mock the useNavigate hook
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

// Mock hooks
jest.mock('@hooks', () => ({
  __esModule: true,
  ...jest.requireActual('@hooks'),
}));

beforeEach(() => {
  // Reset mock function states before each test
  jest.clearAllMocks();
});

describe('EditProduct component', () => {
  it('should render component correctly', () => {
    const { container } = renderWithRouterAndQuery(<EditProduct />);

    expect(container).toBeInTheDocument();
  });

  it('should call handleAddProduct success when submit button is clicked', async () => {
    jest.spyOn(api, 'putData').mockRejectedValue(new Error('Error'));
    (jest.spyOn(hooks, 'useFetchProductDetail') as jest.Mock).mockImplementation(() => ({
      data: MOCK_PRODUCTS[0],
      isFetching: false,
    }));

    const { getByText } = renderWithRouterAndQuery(<EditProduct />);
    const submitBtn = getByText('Submit');

    await act(async () => {
      fireEvent.submit(submitBtn);
    });
  });

  it('should call handleAddProduct error when submit button is clicked', async () => {
    jest.spyOn(api, 'putData').mockResolvedValue(MOCK_PRODUCTS[0]);
    (jest.spyOn(hooks, 'useFetchProductDetail') as jest.Mock).mockImplementation(() => ({
      data: MOCK_PRODUCTS[0],
      isFetching: false,
    }));

    const { getByText } = renderWithRouterAndQuery(<EditProduct />);
    const submitBtn = getByText('Submit');

    await act(async () => {
      fireEvent.submit(submitBtn);
    });
  });
});
