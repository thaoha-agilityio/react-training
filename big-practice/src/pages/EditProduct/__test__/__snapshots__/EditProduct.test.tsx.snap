// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProduct component should render component correctly 1`] = `
<div>
  <div
    class="css-i9nokg"
  >
    <div
      class="chakra-stack css-103yzil"
    >
      <p
        class="chakra-text css-18xzx14"
        data-testid="title"
      >
        Edit Product
      </p>
      <nav
        aria-label="breadcrumb"
        class="chakra-breadcrumb css-0"
      >
        <ol
          class="chakra-breadcrumb__list css-70qvj9"
        >
          <li
            class="chakra-breadcrumb__list-item css-18biwo"
          >
            <a
              class="chakra-breadcrumb__link css-0"
              href="/shop"
            >
              Shop
            </a>
            <span
              class="css-syhbcg"
              role="presentation"
            >
              <svg
                class="chakra-icon css-rozz42"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </li>
          <li
            class="chakra-breadcrumb__list-item css-18biwo"
            data-testid="breadcrumb-item"
          >
            <span
              aria-current="page"
              class="chakra-breadcrumb__link css-0"
            >
              Edit Product
            </span>
          </li>
        </ol>
      </nav>
    </div>
  </div>
  <div
    class="chakra-container css-0"
  >
    <main
      class="css-npgrt2"
    >
      <div
        class="css-xi606m"
      >
        <div
          class="chakra-spinner css-1y7joxr"
        >
          <span
            class="css-8b45rq"
          >
            Loading...
          </span>
        </div>
      </div>
    </main>
  </div>
</div>
`;
